### DISCUSSION TITLE
#define USE_PID compile failed

### DISCUSSION THREAD

**Discussion by dexterbot80** [arendst/Tasmota#19847] [Support request] [answered]
Hello friends, I tried to activte PID function in Tasmota master 13.2.0 and i received this errors.
Can you help me with that?
Thank you
<img width="1145" alt="Screenshot 2023-10-27 082834" src="https://github.com/arendst/Tasmota/assets/47325794/41e9a207-c116-4e9a-908d-b50ee2ebed51">

**Comment by Wintespe** [comment]
I used the Tasmota version 11.1.0.
Furthermore, USE_TIMPROP was activated.
No problems with the Visul Studio Code an PlatformIO

**Comment by dexterbot80** [comment]
> I used the Tasmota version 11.1.0. Furthermore, USE_TIMPROP was activated. No problems with the Visul Studio Code an PlatformIO
Thank you @Wintespe , Can you give an example of settings for a thermostat that operates a hot air heating system?
I want to keep the temperature as linear as possible at 22 degrees

**Comment by Wintespe** [comment]
See: 

https://blog.clanlaw.org.uk/pid-loop-tuning.html

about the PID setup.

<html>
<head>

</head>
<body lang="de-DE" link="#000080" vlink="#800000" dir="ltr">


Command | Def | Description
-- | -- | --
PidPv | - | This Is the current actual 			process value.
PidSp | 19.5 | Setpoint value. This is the 			process value that the process is aiming for.
PidPb | 5 | Proportional band in process 			units (eg degrees). This controls the gain of the loop and is the 			range of process value over which the power output will go from 0 			to full power. The units are that of the process and setpoint, so 			for example in a heating application it might be set to 1.5 			degrees. 			Set 			to zero to implement a 2-step controller.
PidTi | 1800 | This is a setting for the 			integral time, in seconds. It represents the time constant of the 			integration effect. The larger the value the slower the integral 			effect will be. Obviously the slower the process is the larger 			this should be. For example for a domestic room heated by 			convection radiators a setting of one hour might be appropriate 			(in seconds). To disable the integral effect set this to a large 			number,
PidTd | 15 | This is a setting for the 			derivative time, in seconds. It represents the time constant of 			the derivative effect.The larger the value the greater will be the 			derivative effect. Typically this will be set to somewhat less 			than 25% of the integral setting, once the integral has been 			adjusted to the optimum value. To disable the derivative effect 			set this to 0. When initially tuning a loop it is often sensible 			to start with derivative zero and wind it in once other parameters 			have been setup.
PIDInitint | 0.5 | This is an initial value 			which is used to preset the integrated error value when the flow 			is deployed in order to assist in homing in on the setpoint the 			first time. It should be set to an estimate of what the power 			requirement might be in order to maintain the process at the 			setpoint. For example for a domestic room heating application it 			might be set to 0.2 indicating that 20% of the available power 			might be required to maintain the setpoint. The value is of no 			consequence apart from device restart.
PidMaxInterval | 300 | This is 			the maximum time in seconds that is expected between samples. It 			is provided to cope with unusual situations such as a faulty 			sensor that might prevent the node from being supplied with a 			process value. If no new process value is received for 			PidMaxInterval 			then the power is set to the value defined by 			 PidManualPower. 			Set to zerro 			if not used. In this case the controller 			is always 			activated.
PidDSmooth | 3 | IIn 			situations where the process sensor has limited resolution (e.g. 			DS18B20), using the derivative can be problematic when the process 			is changing slowly and the steps in the value cause spikes in the 			derivative. To reduce the effect PidDSmoothr 			can be set to apply a filter to the derivative term. It has been 			found that with the DS18B20 a value of 3 can be beneficial here, 			effectively providing a low pass filter for the derivative at 1/3 			the derivative time. This feature may also be useful if the 			process value is particularly noisy. The smaller the value the 			greater the filtering effect but the more it will reduce the 			effectiveness of the derivative. A value of zero disables this 			feature.
PidAuto | 1 | Auto mode 1 or 0 (for 			manual).This can be used to enable or disable the control 			(1=enable, auto mode, 0=disabled, manual mode). When in manual 			mode the output is set the value defined for PidManualPower
PidManualPower | 0 | Power output when in manual 			mode or fallback mode if too long elapses between process values.
PidUpdateSecs | 0 | Specifies 			the 			interval time in seconds  to 			execute the pid algorithm. 			If 			PidUpdateSecs = 0 the algorithm isonly 			triggerd each time a new temperature value is received. This time 			is specified by TELE_PERIOD. 			A 			value 			of 15 seconds may well be 			appropriate for a domestic room heating application.


</body>
</html>

Make a rule for your settings- Have fun

**Comment by dexterbot80** [comment]
> See:
> 
> https://blog.clanlaw.org.uk/pid-loop-tuning.html
> 
> about the PID setup.
> 
> Command	Def	Description
> PidPv	-	This Is the current actual process value.
> PidSp	19.5	Setpoint value. This is the process value that the process is aiming for.
> PidPb	5	Proportional band in process units (eg degrees). This controls the gain of the loop and is the range of process value over which the power output will go from 0 to full power. The units are that of the process and setpoint, so for example in a heating application it might be set to 1.5 degrees. Set to zero to implement a 2-step controller.
> PidTi	1800	This is a setting for the integral time, in seconds. It represents the time constant of the integration effect. The larger the value the slower the integral effect will be. Obviously the slower the process is the larger this should be. For example for a domestic room heated by convection radiators a setting of one hour might be appropriate (in seconds). To disable the integral effect set this to a large number,
> PidTd	15	This is a setting for the derivative time, in seconds. It represents the time constant of the derivative effect.The larger the value the greater will be the derivative effect. Typically this will be set to somewhat less than 25% of the integral setting, once the integral has been adjusted to the optimum value. To disable the derivative effect set this to 0. When initially tuning a loop it is often sensible to start with derivative zero and wind it in once other parameters have been setup.
> PIDInitint	0.5	This is an initial value which is used to preset the integrated error value when the flow is deployed in order to assist in homing in on the setpoint the first time. It should be set to an estimate of what the power requirement might be in order to maintain the process at the setpoint. For example for a domestic room heating application it might be set to 0.2 indicating that 20% of the available power might be required to maintain the setpoint. The value is of no consequence apart from device restart.
> PidMaxInterval	300	This is the maximum time in seconds that is expected between samples. It is provided to cope with unusual situations such as a faulty sensor that might prevent the node from being supplied with a process value. If no new process value is received for PidMaxInterval then the power is set to the value defined by PidManualPower. Set to zerro if not used. In this case the controller is always activated.
> PidDSmooth	3	IIn situations where the process sensor has limited resolution (e.g. DS18B20), using the derivative can be problematic when the process is changing slowly and the steps in the value cause spikes in the derivative. To reduce the effect PidDSmoothr can be set to apply a filter to the derivative term. It has been found that with the DS18B20 a value of 3 can be beneficial here, effectively providing a low pass filter for the derivative at 1/3 the derivative time. This feature may also be useful if the process value is particularly noisy. The smaller the value the greater the filtering effect but the more it will reduce the effectiveness of the derivative. A value of zero disables this feature.
> PidAuto	1	Auto mode 1 or 0 (for manual).This can be used to enable or disable the control (1=enable, auto mode, 0=disabled, manual mode). When in manual mode the output is set the value defined for PidManualPower
> PidManualPower	0	Power output when in manual mode or fallback mode if too long elapses between process values.
> PidUpdateSecs	0	Specifies the interval time in seconds to execute the pid algorithm. If PidUpdateSecs = 0 the algorithm isonly triggerd each time a new temperature value is received. This time is specified by TELE_PERIOD. A value of 15 seconds may well be appropriate for a domestic room heating application.
> Make a rule for your settings- Have fun



Trust me i use this settings but the thermostat is not working verry well.
For example according my ds18b20 sensor my temperature is 25 degrees celsius and when i set PidSp 27 thermostat relay remain off.

**Comment by Wintespe** [comment]
It took me a while to set the PID controller correctly. At first, I had problems with the power on and off, the thermostat was unstable, which is classic for a control circuit. After that, the result was as follows:

![image](https://github.com/arendst/Tasmota/assets/121858068/d16135e9-c92f-4bcf-a6cb-0cd9ad9b0fe1)

I have switched the setpoint temperature from 25 °C to 30 °C. The time behaviour is now almost perfect.
Another problem could be the TC time.  #17656
At present, this value can only be changed by recompiling, which is a pity.

**Comment by Wintespe** [comment, marked as answer]
Backlog  PidSp 25; PidTi 80; PidTd 25; PidPb 1.2; PidTc 40; PidUpdateSecs 1
TelePeriod 300

Remove PidTc 40, it is not included in the official version of Tasmota..

**Comment by dexterbot80** [comment]
> Backlog PidSp 25; PidTi 80; PidTd 25; PidPb 1.2; PidTc 40; PidUpdateSecs 1 TelePeriod 300
> 
> Remove PidTc 40, it is not included in the official version of Tasmota..

It's the first time the thermostat moves as it should :)
What does that PidTc 40 do?

**Comment by Wintespe** [comment]
See #17656 Motivation
PidTc 40 set the time to 40 sec
